### PR TITLE
Refactor EffPlaySound to be more readable and resistant to versioning issues.

### PR DESCRIPTION
**Pull request by sovdeeth**
### Description

The current state of EffPlaySound relies on a lot of nested FunctionalInterfaces with static methods that makes it quite hard to follow, read, and debug. Issues with differing version support has resulted in two bugs so far, one that has caused an inability to use entity emitters on 1.18 despite the API existing, since the current impl isn't flexible enough to allow it.

This PR re-introduces that feature by refactoring EffPlaySound to use an adapter pattern to unify Player and World playSound methods. It also moves sound validation, among other things, outside of the player loop which should both improve performance and fix an issue where the sound expression was being evaluated once for each player.

I have done manual testing on 1.17, 1.18, and 1.21, though I think junit tests should be implemented via mocking, if feasible. I have not yet looked into that. I would also appreciate any users with existing sound-heavy code to test this PR's nightly out if they can (click Checks -> Java 21 CI -> download from Artifacts).

---
**Target Minecraft Versions:** any 
**Requirements:** none 
**Related Issues:** #7016, #6907, #6910

### CODE REVIEW

**Review by UnderscoreTud — APPROVED** [review]

**Review by Moderocky — APPROVED** [review]
Good.

File: src/main/java/ch/njol/skript/bukkitutil/sounds/SoundReceiver.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+ * Adapter pattern to unify {@link World} and {@link Player} playSound methods.
+ * Methods can be called without determining version support, it is handled internally.
+ * Non-supported methods will simply delegate to supported methods.
+ */
+public interface SoundReceiver {
+
+	boolean ADVENTURE_API = Skript.classExists("net.kyori.adventure.sound.Sound$Builder");
+	boolean SPIGOT_SOUND_SEED = Skript.methodExists(Player.class, "playSound", Entity.class, Sound.class, SoundCategory.class, float.class, float.class, long.class);
+	boolean ENTITY_EMITTER_SOUND = Skript.methodExists(Player.class, "playSound", Entity.class, Sound.class, SoundCategory.class, float.class, float.class);
+	boolean ENTITY_EMITTER_STRING = Skript.methodExists(Player.class, "playSound", Entity.class, String.class, SoundCategory.class, float.class, float.class);
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/bukkitutil/sounds/SoundReceiver.java:25]: Static so multiple extending interfaces don't reinitialize heavy reflection checks.
```suggestion
	static boolean ADVENTURE_API = Skript.classExists("net.kyori.adventure.sound.Sound$Builder");
	static boolean SPIGOT_SOUND_SEED = Skript.methodExists(Player.class, "playSound", Entity.class, Sound.class, SoundCategory.class, float.class, float.class, long.class);
	static boolean ENTITY_EMITTER_SOUND = Skript.methodExists(Player.class, "playSound", Entity.class, Sound.class, SoundCategory.class, float.class, float.class);
	static boolean ENTITY_EMITTER_STRING = Skript.methodExists(Player.class, "playSound", Entity.class, String.class, SoundCategory.class, float.class, float.class);
```

File: src/main/java/ch/njol/skript/bukkitutil/sounds/AdventureSoundUtils.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+import org.bukkit.NamespacedKey;
+import org.bukkit.SoundCategory;
+import org.bukkit.World;
+import org.bukkit.entity.Entity;
+import org.bukkit.entity.Player;
+
+import java.util.OptionalLong;
+
+public class AdventureSoundUtils {
+	public static Sound getAdventureSound(NamespacedKey key, SoundCategory category, float volume, float pitch, OptionalLong seed) {
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/bukkitutil/sounds/AdventureSoundUtils.java:14]:
```suggestion
public class AdventureSoundUtils {

	public static Sound getAdventureSound(NamespacedKey key, SoundCategory category, float volume, float pitch, OptionalLong seed) {
```

File: src/main/java/ch/njol/skript/bukkitutil/sounds/SoundReceiver.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+				this.playSound(entity, sound.getKey(), category, volume, pitch);
+			} else {
+				player.playSound(entity, sound.getKey(), category, volume, pitch, seed.getAsLong());
+			}
+		}
+	}
+
+	// World adapter pattern
+	class WorldSoundReceiver implements SoundReceiver {
+		private final World world;
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/bukkitutil/sounds/SoundReceiver.java:85]:
```suggestion
	class WorldSoundReceiver implements SoundReceiver {
	
		private final World world;
```

**Review by TheLimeGlass — APPROVED** [review]
I still find kiip's FunctionalInterface usage here easier to read.

Performance was the same for execution. Initialization of the EffPlaySound class was 50-90ms better than existing.

File: src/main/java/ch/njol/skript/bukkitutil/sounds/SoundReceiver.java (hunk 1 above)
Comment (sovdeeth) [reply src/main/java/ch/njol/skript/bukkitutil/sounds/SoundReceiver.java:25]: They already are static, you can't have non-static interface fields.

Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/bukkitutil/sounds/SoundReceiver.java:25]: The exact reason to keep the static definition

Comment (sovdeeth) [reply src/main/java/ch/njol/skript/bukkitutil/sounds/SoundReceiver.java:25]: lime adding static doesn't do anything but pop up an ide inspection saying that you should remove it. there's no point to adding the keyword static here.
